existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,160 @@
+The syntax is very simple - at the beginning of function or instance declaration, before any constraints or arguments were introduced use
+quantifier with all type variables you want to introduce later (Note: you get a compilation error if you use undeclared
+type variables after forall):
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+Here we introduce four variables. We might want to introduce some constraints.
+For example, it is very likely from the function definition that `m` is an instance of `Monad` typeclass.
+In order to add this constraint we need to place it after all variables are declared:
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```  
+You might want to ask, what is so special about this syntax?
+Nothing changed since we added the quantifier.
+Well, actually there are few things to notice here.
+
+##### Order of type variables
+First of all, now you can change the order of type variables.
+Let's imagine you want to define function with 10 type variables (Yeah, there are real life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.
+Without forall you will write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```  
+
+Quite long right? 
+This is because ghc will maintain the same order, as these variables first appear in function definition.
+However, using explicit `forall.` this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.
+
+##### Additional type variables
+Second, you now can define functions with more type variables than you use on the right side of declaration.
+This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:
+
+``` Haskell
+someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
+```
+ 
+Here we don't have `env` used in the type definition, but despite that, we still able to add `MonadReader` constraint.
+
+## Existential quantification
+Beside universally quantified variables, Haskell also supports existential types. 
+In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
+That might be a little confusing.
+The following construction is called existential type in Haskell - when you use forall quantifier in the right-hand side of function or datatype declaration. 
```

Comment:
I would explain here why it results in an existential type, and how it relates to existential quantification, think that would be interesting.